### PR TITLE
check label for invalid code points

### LINKED ISSUE #227: Detect invalid ACE?
Hi,

I would like to know if in this library is there any way to detect an invalid domain ACE format when using the idn package? (FromPunycode)

For example, the FQDN **xn--7warambrecmdctfmhlrtrmbpyjdccaaahkrgl4chlip9bclcgddbb4ooioa.bnnhg** according to [Verisign](http://mct.verisign-grs.com/convertServlet?input=xn--7warambrecmdctfmhlrtrmbpyjdccaaahkrgl4chlip9bclcgddbb4ooioa.bnnhg) fails due to `Error:: Disallowed/Unassigned 3A3`, but the FromPunycode returns `σειράτάξησυπουργείωνΣύνθεσηυπουργικούσυμβουλίουουουο.bnnhg`.

```go
package main

import (
	"github.com/miekg/dns/idn"
)

func main() {
	output := idn.FromPunycode("xn--7warambrecmdctfmhlrtrmbpyjdccaaahkrgl4chlip9bclcgddbb4ooioa.bnnhg")
	println(output)
}
```

Best regards,
Rafael Justo

### PR DESCRIPTION
**Pull request by rafaeljusto**
Closes #227

### CODE REVIEW

**Comment by miekg** [comment]
cool. Can you extract the logic from idna_table.go into punycode.go, so only the table and types remain there? Stupid github won't let me commit on the file (some small naming nitpicks)

**Comment by rafaeljusto** [comment]
done!

File: idn/punycode.go [hunk 1]
```diff
@@ -267,3 +272,32 @@ func decode(b []byte) []byte {
 	}
 	return ret.Bytes()
 }
+
+// isValidRune checks if the character is valid. We will look for the
+// character property in the IDNA table. For now we aren't checking special
+// rules in case of contextual property
+func isValidRune(r rune) bool {
+	return findProperty(r) == idnaPropertyPVALID
```
Comment (miekg) [inline idn/punycode.go]: is 'idna' really important? we are already in the idn package, so dropping 'idna' makes sense?

File: idn/punycode.go [hunk 2]
```diff
@@ -267,3 +272,32 @@ func decode(b []byte) []byte {
...
+func findProperty(r rune) idnaProperty {
+	imin, imax := 0, len(idnaTable)
+
+	for imax >= imin {
+		imid := (imin + imax) / 2
+
+		codePoint := idnaTable[imid]
+		if (codePoint.start == r && codePoint.end == 0) || (codePoint.start <= r && codePoint.end >= r) {
+			return codePoint.state
+		} else if (codePoint.end > 0 && codePoint.end < r) || (codePoint.end == 0 && codePoint.start < r) {
```
Comment (miekg) [inline idn/punycode.go]: as you return above, you can just start a new if (no need for else if)

**Comment by rafaeljusto** [comment]
Thanks for the code review, I already made the suggested changes.

**Comment by miekg** [comment]
cool and thanks!

**Pull request merged**